### ISSUE TITLE
UrlAttribute and EmailAttribute validation with Blazor does not allow empty values

### ISSUE THREAD

**Issue by Eirenarch** [dotnet/runtime#53820]
When trying to validate with Url and Email attributes in Blazor there is some weird behavior. When the value is optional you can't delete it. If the input is empty you can submit the form (as you should) but if you type something and delete it or if it had value (say you are editing existing data) and you delete it the validation fails because the validation kicks in despite the fact that the value is optional. This behavior is inconsistent internally (it shouldn't matter if the field was empty or the data was deleted) and also inconsistent with the behavior in the attributes elsewhere for example in ASP.NET MVC views

### To Reproduce
Using the default Blazor WASM template in .NET 5 add the following page

```
@page "/url"
@using TestBlazorApp.Shared

<h1>URL</h1>

<EditForm Model="@Model">
    <DataAnnotationsValidator />
    <ValidationSummary />

    URL:
    <InputText @bind-Value="Model.Url" />

    <br />

    URL Regex:
    <InputText @bind-Value="Model.UrlRegex" />

    <br />

    Email:
    <InputText @bind-Value="Model.Email" />
    
    <br />
    <button type="submit">Save</button>
</EditForm>
@code {

    UrlModel Model = new UrlModel();

}
```

And the following class in the shared project

```
public class UrlModel
{
    [Url]
    public string? Url { get; set; }

    [RegularExpression(@"https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)")]
    public string? UrlRegex { get; set; }

    [EmailAddress]
    public string? Email { get; set; }
}
```

To reproduce write something in the fields move to another field come back and delete the value.

Result:

![image](https://user-images.githubusercontent.com/2217874/120848996-2d360c00-c57e-11eb-9fc0-3843891a3a4e.png)

I am testing with Blazor WASM in .NET 5 with Microsoft Edge but I believe this happens with Blazor Server, Blazor Client and all versions of .NET and all browsers

**Comment by Eirenarch** [comment]
@mkArtakMSFT I don't think this is a problem with the validation logic. It is a problem with the fact that an empty input in Blazor results in two different values bound to the model. Sometimes it is null, sometimes it is empty string

**Comment by SteveSandersonMS** [comment]
@Eirenarch It's pretty standard across UI frameworks to represent empty textboxes as having a value of "empty string". If Blazor was to normalize this to `null` that would be surprising to more people. So unfortunately we couldn't really do that kind of normalization by default. Arguably we could auto-write an empty string over any `null`-valued property as soon as you bind it to a textbox, but (1) that would trigger setters that might have side-effects, which in turn would surprise and inconvenience some people, and (2) you'd still have the same problem with `[Url]` that you're reporting here (i.e., empty string treated as invalid despite not being required).

So overall I don't think we have much of a choice to do something different by default. If you really want, you can use custom `get`/`set` logic to normalize an empty string to a null value, though I appreciate that's inconvenient.

**Comment by nsimeonov** [comment]
How about RequiredFieldAttribute as @Grauenwolf pointed? Validation attributes should ignore empty values unless the field is tagged as [Required]. Null or not - empty value is ok on any field, regardless of it's other validation attributes, if it's not required.

**Comment by SteveSandersonMS** [comment]
@nsimeonov I agree. That's why we moved this issue to the `dotnet/runtime` repo which owns the code for `UrlAttribute` etc.

**Comment by Eirenarch** [comment]
@SteveSandersonMS I'd argue that the expected behavior here is the one of the rest of ASP.NET where empty textboxes result in null value. Sure that might be an artefact of the way forms on the web work but it has been this way in ASP.NET for decades and is true for MVC and Razor Pages. In my opinion it would be wise to normalize empty string to null in the component itself.

**Comment by ajcvickers** [comment]
Note from triage: we are not going to change the default behavior of existing validation attributes because such changes are very likely to be breaking for some consumers. It's also problematic to change these behaviors because there is code out there that reads the annotations and reacts appropriately, rather than using the built-in validation. This code would then diverge from the behavior implemented in the attribute unless the consumer is updated, and many of the consumers are unlikely to ever be updated.

The bottom line here is that the validation attributes are what they are. It's old and not very good code. I would not recommend it going forward. At some point there probably needs to be a new validation library that is easier to evolve and maintain, but there are currently no concrete plans for this.

**Comment by Eirenarch** [comment]
But this is not an issue with the attributes. This is an issue with the fact that Blazor thinks empty input is empty string unlike everything else in ASP.NET which treats empty input as null.
